tests/health_checks/NDv4/test-bandwidth.sh

```diff
@@ -35,13 +35,251 @@ catch_error() {
+inline void gpuStatus(cudaError_t status, const char *file, int line)
+{
+    if (status != cudaSuccess) 
+    {	
+        fprintf(stderr,"GPUassert: %s %s %d\n", cudaGetErrorString(status),\
+			file, line);
+        exit(status);
+    }
+}
+
+
+__global__ void dummy_kernel(uint8_t * dummy_var)
+{
+}
+
+__global__ void initialize_ker(const size_t numEntries, uint8_t *gpuBuffer)
+{
+    uint64_t index = blockIdx.x * blockDim.x  + threadIdx.x;
+    if ( index < numEntries)
+    {
+        gpuBuffer[index] = index % 256;
+    }
+}
+
+
+inline void initialize(const int num_numa, const int num_gpus,\
+		const size_t numEntries, uint8_t **gpuBuffer,\
+		uint8_t **hostBuffer)
+{
+    const size_t blocksize = 256;
+    const size_t numBlocks = (numEntries + blocksize - 1) / blocksize;
+
+    const size_t size = numEntries * sizeof(uint8_t);
+
+    for(int gpu=0;gpu<num_gpus;++gpu)
+    {
+        gpuErrorCheck(cudaSetDevice(gpu));
+        gpuErrorCheck(cudaMalloc( (void **) &(gpuBuffer[gpu]), size));
+
+        initialize_ker<<<numBlocks, blocksize>>>(numEntries,gpuBuffer[gpu]);
+    }
+    for(int numa=0;numa<num_numa;++numa)
+    {
+        hostBuffer[numa] =static_cast<uint8_t *>(numa_alloc_onnode(size,numa));
+        gpuErrorCheck(cudaHostRegister(hostBuffer[numa], size,\
+				cudaHostRegisterMapped));
+        cudaMemcpy(hostBuffer[numa],gpuBuffer[0],numEntries*sizeof(uint8_t),\
+			cudaMemcpyDefault);
+        cudaDeviceSynchronize();
+    }
+}
+
+inline void cleanup(const int num_numa, const int num_gpus,\
+		const size_t numEntries, uint8_t **gpuBuffer,\
+		uint8_t **hostBuffer)
+{
+    const size_t size = numEntries * sizeof(uint8_t);
+
+
+    for(int gpu=0;gpu<num_gpus;++gpu)
+    {
+        gpuErrorCheck(cudaFree(gpuBuffer[gpu]));
+    }
+    for(int numa=0;numa<num_numa;++numa)
+    {
+        numa_free(hostBuffer[numa], size);
+    }
+
+    delete[] hostBuffer;
+    delete[] gpuBuffer;
+}
+
+inline void copy_loop(const int num_gpus,const int num_numa,const int warmup,\
+		const int loops, const size_t numEntries, const bool time,\
+		uint8_t ** gpuBuffer, uint8_t **hostBuffer, const int htod)
+{
+    const int size = numEntries * sizeof(uint8_t);
+    for(int gpu=0;gpu<num_gpus;++gpu)
+    {
+        gpuErrorCheck(cudaSetDevice(gpu));
+        for(int numa=0;numa<num_numa;++numa)
+        {
+    	    cudaEvent_t start,stop;
+    	    cudaEventCreate(&start);
+    	    cudaEventCreate(&stop);
+            double totaltime = 0.0;
+    	    float localtime = 0.0;
+    	    for(int iloop=0;iloop<warmup+loops;++iloop)
+    	    {
+    	        cudaEventRecord(start,0);
+    		if(htod){gpuErrorCheck(cudaMemcpyAsync(gpuBuffer[gpu],\
+				hostBuffer[numa],size,cudaMemcpyDefault));}
+    		else{gpuErrorCheck(cudaMemcpyAsync(hostBuffer[numa],\
+				gpuBuffer[gpu],size,cudaMemcpyDefault));}
+    	        cudaEventRecord(stop,0);
+    	        cudaEventSynchronize(stop);
+    
+    	        dummy_kernel<<<1, 64>>>(gpuBuffer[gpu]);
```

Comment:
The dummy kernel isn't necessary, I just added it out of paranoia that maybe a compiler could get the idea to optimize out the memory copies if it can see that I'm not doing anything with the data. Good catch, I should delete it or add a comment at least...

Reply:
Discussed off line. marking as resolved